src/Utilities/MoistThermodynamics/src/MoistThermodynamics.jl

```diff
@@ -7,504 +7,951 @@ saturation specific humidities
 """
-function latent_heat_generic(T, LH_0, cp_diff)
+latent_heat_generic(T::Real, LH_0::Real, Δcp::Real) =
+  LH_0 + Δcp * (T - T_0)
 
-    return LH_0 + cp_diff * (T - T_0)
 
-end
+"""
+    Phase
 
+A phase condensate, to dispatch over
+[`saturation_vapor_pressure`](@ref) and
+[`saturation_shum_generic`](@ref).
+"""
 abstract type Phase end
+
+"""
+    Liquid <: Phase
+
+A liquid phase, to dispatch over
+[`saturation_vapor_pressure`](@ref) and
+[`saturation_shum_generic`](@ref).
+"""
 struct Liquid <: Phase end
+
+"""
+    Ice <: Phase
+
+An ice phase, to dispatch over
+[`saturation_vapor_pressure`](@ref) and
+[`saturation_shum_generic`](@ref).
+"""
 struct Ice <: Phase end
 
 """
-    `saturation_vapor_pressure(T, Liquid())`
+    saturation_vapor_pressure(T, Liquid())
+
 Return the saturation vapor pressure over a plane liquid surface at
 temperature `T`.
 
-    `saturation_vapor_pressure(T, Ice())`
+    saturation_vapor_pressure(T, Ice())
+
 Return the saturation vapor pressure over a plane ice surface at
 temperature `T`.
 
-    `saturation_vapor_pressure(T, LH_0, cp_diff)`
+    saturation_vapor_pressure(T, LH_0, Δcp)
 
 Compute the saturation vapor pressure over a plane surface by integration
 of the Clausius-Clepeyron relation.
 
 The Clausius-Clapeyron relation
 
-    dlog(p_vs)/dT = [LH_0 + cp_diff * (T-T_0)]/(R_v*T^2)
+    dlog(p_v_sat)/dT = [LH_0 + Δcp * (T-T_0)]/(R_v*T^2)
 
 is integrated from the triple point temperature `T_triple`, using
 Kirchhoff's relation
 
-    L = LH_0 + cp_diff * (T - T_0)
+    L = LH_0 + Δcp * (T - T_0)
 
 for the specific latent heat `L` with constant isobaric specific
 heats of the phases. The linear dependence of the specific latent heat
 on temperature `T` allows analytic integration of the Clausius-Clapeyron
-relation to obtain the saturation vapor pressure `p_vs` as a function of
+relation to obtain the saturation vapor pressure `p_v_sat` as a function of
 the triple point pressure `press_triple`.
 
 """
-saturation_vapor_pressure(T, ::Liquid) = saturation_vapor_pressure(T, LH_v0, cp_v - cp_l)
-saturation_vapor_pressure(T, ::Ice) = saturation_vapor_pressure(T, LH_s0, cp_v - cp_i)
+saturation_vapor_pressure(T::DT, ::Liquid) where DT = saturation_vapor_pressure(T, DT(LH_v0), DT(cp_v) - DT(cp_l))
+function saturation_vapor_pressure(ts::ThermodynamicState, ::Liquid)
+
+    return saturation_vapor_pressure(air_temperature(ts), LH_v0, cp_v - cp_l)
 
-function saturation_vapor_pressure(T, LH_0, cp_diff)
+end
+saturation_vapor_pressure(T::DT, ::Ice) where DT = saturation_vapor_pressure(T, DT(LH_s0), DT(cp_v) - DT(cp_i))
+saturation_vapor_pressure(ts::ThermodynamicState, ::Ice) = saturation_vapor_pressure(air_temperature(ts), LH_s0, cp_v - cp_i)
+
+function saturation_vapor_pressure(T::DT, LH_0::DT, Δcp::DT) where DT
 
-    return press_triple * (T/T_triple)^(cp_diff/R_v) *
-        exp( (LH_0 - cp_diff*T_0)/R_v * (1 / T_triple - 1 / T) )
+    return DT(press_triple) * (T/DT(T_triple))^(Δcp/DT(R_v)) *
```

Comment:
It seems like `T/DT(T_triple)` is negative?

Reply:
Yes, temperatures returned from saturation adjustment were as low as ~`0.001`. I suspect that there is an error in the isentropic vortex test.

The test is breaking [here](https://github.com/climate-machine/CLIMA/blob/master/src/DGmethods/test/Euler/isentropic_vortex.jl#L55), but I'm suspicious of [this](https://github.com/climate-machine/CLIMA/blob/master/src/DGmethods/test/Euler/isentropic_vortex.jl#L54) equation. I don't see why internal energy at zero Kelvin is added to the RHS, this doesn't seem to match the design docs. @jkozdon, I tried adjusting this line, but it didn't seem to fix the test (sorry for roping you into `MoistThermodynamics.jl`!).

Reply:
For now, I've removed `isentropic_vortex` from the test set. We can add it back in once we figure out what's wrong.

Reply:
@jkozdon Yes, the initialization of energy in this [line](https://github.com/climate-machine/CLIMA/blob/5a2b2eee73da88629b3472b0b7c11c6857542873/src/DGmethods/test/Euler/isentropic_vortex.jl#L113) is not consistent with the thermodynamics. That line in isentropic_vortex.jl needs to be replaced by a call to `total_energy`, with temperature from `air_temperature`. The `internal_energy(0)` in [this](https://github.com/climate-machine/CLIMA/blob/5a2b2eee73da88629b3472b0b7c11c6857542873/src/DGmethods/test/Euler/isentropic_vortex.jl#L54) then needs to be removed.

Reply:
Sorry for missing this conversation.

I based this based on what was in the current Dycore code for [`isentropic_vortex.jl`](https://github.com/climate-machine/CLIMA/blob/d13e455e315cea579c69263af02f5ca562a6af79/src/Atmos/Dycore/drivers/isentropic_vortex.jl#L68) which came in with Moist thermo pr #138. 

The movement of `internal_energy(0)` out of the initial condition was done to be more `Float32` friendly.

For now I am deleting the test from the balance law complete. When the new framework is in place proper integration tests with thermodynamics can be added.